src/nc_sentinel.c

```diff
@@ -0,0 +1,471 @@
+
+    sentinel->owner->next_sentinel_connect = 0LL;
+
+    return NC_OK;
+}
+
+void
+sentinel_close(struct context *ctx, struct conn *conn)
+{
+    struct server_pool *pool;
+
+    pool = ((struct server*)conn->owner)->owner;
+
+    sentinel_set_next_connect(pool);
+
+    conn->status = CONN_DISCONNECTED;
+
+    server_close(ctx, conn);
+}
+
+static rstatus_t
+sentinel_proc_sentinel_info(struct context *ctx, struct server *sentinel, struct msg *msg)
+{
+    rstatus_t status;
+    int i, master_num, switch_num, server_port;
+    struct string server_name, server_ip, sentinel_masters_prefix, master0_info_prefix, tmp_string;
+    struct server *server;
+    struct mbuf *line_buf;
+
+    string_init(&tmp_string);
+    string_init(&server_name);
+    string_init(&server_ip);
+    string_set_text(&sentinel_masters_prefix, "sentinel_masters");
+    string_set_text(&master0_info_prefix, "master0");
+
+    line_buf = mbuf_get();
+    if (line_buf == NULL) {
+        goto error;
+    }
+
+    /* get sentinel master num at line 3 */
+    msg_read_line(msg, line_buf, 3);
+    if (mbuf_length(line_buf) == 0) {
+        log_error("read line failed from sentinel ack info when skip line not used.");
+        goto error;
+    }
+    status = mbuf_read_string(line_buf, ':', &tmp_string);
+    if (status != NC_OK || string_compare(&sentinel_masters_prefix, &tmp_string)) {
+        goto error;
+    }
+    status = mbuf_read_string(line_buf, CR, &tmp_string);
+    if (status != NC_OK) {
+        goto error;
+    }
+    master_num = nc_atoi(tmp_string.data, tmp_string.len);
+    if (master_num < 0) {
+        log_error("parse master number from sentinel ack info failed.");
+        goto error;
+    }
+
+    /* parse master info from sentinel ack info */
+    i = 0;
+    switch_num = 0;
+    while (i < master_num) {
+        msg_read_line(msg, line_buf, 1);
+        if (mbuf_length(line_buf) == 0) {
+            log_error("read line failed from sentinel ack info when parse master item.");
+            goto error;
+        }
+
+        /* skip some useless lines before the master info */
+        if (i == 0) {
+            status = mbuf_read_string(line_buf, ':', &tmp_string);
+            if (status != NC_OK || string_compare(&master0_info_prefix, &tmp_string)) {
+                continue;
+            }
+        }
+
+        /* skip master item server name prefix */
+        status = mbuf_read_string(line_buf, '=', NULL);
+        if (status != NC_OK) {
+            log_error("skip server name prefix failed.");
+            goto error;
+        }
+
+        /* get server name */
+        status = mbuf_read_string(line_buf, ',', &server_name);
+        if (status != NC_OK) {
+            log_error("get server name failed.");
+            goto error;
+        }
+
+        server = server_find_by_name(ctx, sentinel->owner, &server_name);
+        if (server == NULL) {
+            log_error("unknown server name:%.*s", server_name.len, server_name.data);
```

Comment:
This should just be a continue and info/warning log. A sentinel server may refer to multiple redis servers, and the twemproxy config may only be a subset of them.

Developers might also temporarily set up additional servers with a given sentinel cluster to test.

Also, the sentinel patches were applied to the fork https://github.com/ifwe/twemproxy - which I'd planned to merge once I'd created a release for all of the other non-sentinel patches, though that's been delayed, and I expect resolving merge conflicts might take a while.

I should create a release - there were some edge cases I was trying to track down months ago, but I believe they weren't new edge cases caused by the latest patches